### ISSUE TITLE
Did not catch non return value on a function that should return a value.

### ISSUE THREAD

**Issue by Huixiong91** [Microsoft/TypeScript#19706]
Given the following function definition, complier did not complain about the possible `if` block being executed, which we did not return any value;
```TypeScript
function someFunc(params: any): Promise<any> {
  if (!params) {
    // we didn't return any Promise, why didn't complier catch & complain about it?
    throw new Error('Error thrown') 
  }
  return new Promise((resolve, reject) => {
    resolve("Success")
  });
}
```
Someone could have make a call passing in a `null` value, something like this "`someFunc(null)`" expecting to get a Promise back, only to be surprised that nothing was returned?

**Comment by Huixiong91** [comment]
Sorry, I realise this is not a bug at all, I was throwing an error.

The code was properly meant to do something like this. Throwing an error inside the `Promise` executor block.
```TypeScript
function someFunc(params: any): Promise<any> {

  return new Promise((resolve, reject) => {
      if (!params) { throw new Error('Error thrown') }
      resolve("Success")
  });
}
```
